### ISSUE TITLE
[Dev] Crash on WorldSocket::ReadHeaderHandler

### ISSUE THREAD

**Issue by Epicurus4** [TrinityCore/TrinityCore#14944]
Branch 6.x
Using methods from commit 812fa65c613ecaa603dafc708df0088f3b50a90e

Crash log can be found [here](https://gist.github.com/Epicurus4/dded2254aec6e09d0af8).

**Comment by Epicurus4** [comment]
Branch 6.x
Commit 812fa65c613ecaa603dafc708df0088f3b50a90e

Update with related crash:
https://gist.github.com/Epicurus4/d516f2c0dff4d35280b9

**Comment by DDuarte** [comment]
The assert macro doesn't seem to be doing its job.

```
    ASSERT(_headerBuffer.GetActiveSize() == SizeOfClientHeader[_initialized][_authCrypt.IsInitialized()], "Header size " SZFMTD " different than expected %u", _headerBuffer.GetActiveSize(), SizeOfClientHeader[_initialized][_authCrypt.IsInitialized()]);
```

**Comment by Shauren** [comment]
@DDuarte it prints to stderr, not into gdb logging

**Comment by BrettDioson** [comment]
I believe I am also getting this error right after authenticating and just before the character screen is brought up (Retrieving Characters):

P:\Downloads\Games\WoW\Trinity6x\Source\src\server\game\Server\WorldSocket.cpp:246 in WorldSocket::ReadHeaderHandler ASSERTION FAILED:  _headerBuffer.GetActiveSize() == SizeOfClientHeader[_initialized][_authCrypt.IsInitialized()] Header size 4 different than expected 6

Branch 6.x
Commit https://github.com/TrinityCore/TrinityCore/commit/a177a10018fb081c5a29f13e5e473be2f8a5316e

I don't know how widespread this is.

**Comment by tkrokli** [comment]
One of my contacts got the same error as @BrettDioson, it looks like this:
```
C:\Users\surafbrov\Downloads\TrinityCore\6.x\src\server\game\Server\WorldSocket.cpp:246 in WorldSocket::ReadHeaderHandler
ASSERTION FAILED:
  _headerBuffer.GetActiveSize() == SizeOfClientHeader[_initialized][_authCrypt.IsInitialized()] Header size 4 different than expected 6 
```
This was on f44364d though (build 20253), it also caused a Windows crash with error 0xC0000005.

**Comment by BrettDioson** [comment]
@tkrokli , out of curiosity does your contact need to start the patched wow exe with the uid to get it to work?:

       wow-64_Patched.exe -uid wow_enus

That's the only way I have been able to start my wow client.  Trying to see if we have some similarities.  I've tried clearing my cache, but I have no way of getting into the core without it crashing until this is fixed.  I have tried with both the 64bit and 32bit executables (patched).

**Comment by BrettDioson** [comment]
Adding crashlog for my latest build.

Branch: 6x
Commit: https://github.com/TrinityCore/TrinityCore/commit/75bc726c9bedf7cb736ff8ef96d30084bac226be
Crashlog https://gist.github.com/BrettDioson/1d34b2580efc1b59e980

**Comment by joschiwald** [comment]
@BrettDioson crashlogs in release mode are useless, you have to build in debug mode

**Comment by BrettDioson** [comment]
@joschiwald Updated.  sorry about that.  Please let me know if I did it correctly.

**Comment by BrettDioson** [comment]
Created an additional crashlog from build with additional core debug:

Branch: 6x
Commit:  c35e323bc916cf791fba7157a1f749a862e54494
Crashlog: https://gist.github.com/BrettDioson/7214e97770e5f0d7e3a6

Error:
P:\Downloads\Games\WoW\Trinity6x\Source\src\server\game\Server\WorldSocket.cpp:246 in WorldSocket::ReadHeaderHandler ASSERTION FAILED:
  _headerBuffer.GetActiveSize() == SizeOfClientHeader[_initialized][_authCrypt.IsInitialized()] Header size 4 different than expected 6
WorldSocket::HandleAuthSession: Client 1#1 (xxx.xxx.xxx.xxx) attempted to log in using deprecated login method (GRUNT).
(0) : Run-Time Check Failure #2 - Stack around the variable 'buffer' was corrupted.(0) : Run-Time Check Failure #2 - Stack around the variable 'buffer' was corrupted.(0) : Run-Time Check Failure #2 - Stack around the variable 'buffer' was corrupted.(0) : Run-Time Check Failure #2 - Stack around the variable 'buffer' was corrupted.

Windows Server 2008 Hyper-V Guest 6core/12GB
Built with cMake3.2.3/Vs2013/Boots1.58/oSSL1.0.2d/zMQ4.0.4

**Comment by Razinao** [comment]
I have had several people encounter this issue and the solution thus far has been to get them to navigate to the Battlenet folder inside ProgramData\Blizzard Entertainment and rename it to something else, and then get them to run the connection_patcher which will download the appropriate modules before patching them.

**Comment by BrettDioson** [comment]
Thanks @Razinao.  I tried that, actually I have also done a full delete/clean reg, and reinstall of the full wow client and battle.net.  I have the same issue.

**Comment by Razinao** [comment]
Very well. Whilst I was trying (incorrectly) to find out what the issue was caused by, I went and created a new database and ran the base/characters_database.sql on it, then changed all references to the old character database to the new one. It seemed to solve the issue after the solution I gave out above, but when I migrated my old characters database over the newly created one the issue started happening again.
